src/lib/structures/Josh.ts

```diff
@@ -0,0 +1,1813 @@
+import { Awaitable, isFunction, isPrimitive, Primitive } from '@sapphire/utilities';
+import { writeFile } from 'fs/promises';
+import { emitWarning } from 'process';
+import type { CoreMiddleware as AutoEnsureMiddleware } from '../../middlewares/CoreAutoEnsure';
+import { JoshError } from '../errors';
+import { convertLegacyExportJSON, isNodeEnvironment } from '../functions';
+import {
+  AutoKeyPayload,
+  ClearPayload,
+  DecPayload,
+  DeletePayload,
+  EnsurePayload,
+  EveryHook,
+  EveryPayload,
+  FilterHook,
+  FilterPayload,
+  FindHook,
+  FindPayload,
+  GetAllPayload,
+  GetManyPayload,
+  GetPayload,
+  HasPayload,
+  IncPayload,
+  KeysPayload,
+  MapHook,
+  MapPayload,
+  MathPayload,
+  PartitionHook,
+  PartitionPayload,
+  Payload,
+  PushPayload,
+  RandomKeyPayload,
+  RandomPayload,
+  RemoveHook,
+  RemovePayload,
+  SetManyPayload,
+  SetPayload,
+  SizePayload,
+  SomeHook,
+  SomePayload,
+  UpdateHook,
+  UpdatePayload,
+  ValuesPayload
+} from '../payloads';
+import { BuiltInMiddleware, KeyPath, KeyPathJSON, MathOperator, Method, Path, StringArray, Trigger } from '../types';
+import { isLegacyExportJSON } from '../validators';
+import { MapProvider } from './default-provider';
+import { JoshProvider } from './JoshProvider';
+import { Middleware } from './Middleware';
+import { MiddlewareStore } from './MiddlewareStore';
+
+/**
+ * The base class that makes Josh work.
+ * @see {@link Josh.Options} for all options available to the Josh class.
+ * @since 2.0.0
+ *
+ * @example
+ * ```typescript
+ * const josh = new Josh({
+ *   name: 'name',
+ *   // More options...
```

Comment:
does typescript not have the `provider` option? :P

Reply:
It does, but I never bothered to put it here because it isn't a required property anymore unlike `name`, but I can see why it should be here as an example. The main reason I didn't keep it here is because it just shows how quickly you can create a Josh instance and use it.